apollo-router/src/plugins/cache/entity.rs

```diff
@@ -0,0 +1,618 @@
+
+            Ok(ControlFlow::Break(
+                subgraph::Response::builder()
+                    .data(value.0.data)
+                    .extensions(Object::new())
+                    .context(request.context)
+                    .build(),
+            ))
+        }
+        None => {
+            request
+                .context
+                .private_entries
+                .lock()
+                .insert(RootCacheKey(key));
+
+            Ok(ControlFlow::Continue(request))
+        }
+    }
+}
+
+struct EntityCacheResults(Vec<IntermediateResult>);
+
+async fn cache_lookup_entities(
+    name: String,
+    cache: RedisCacheStorage,
+    mut request: subgraph::Request,
+) -> Result<ControlFlow<subgraph::Response, subgraph::Request>, BoxError> {
+    let body = request.subgraph_request.body_mut();
+
+    let keys = extract_cache_keys(&name, body, &request.context)?;
+
+    let cache_result: Vec<Option<CacheEntry>> = cache
+        .get_multiple(keys.iter().map(|k| RedisKey(k.clone())).collect::<Vec<_>>())
+        .await
+        .map(|res| res.into_iter().map(|r| r.map(|v| v.0)).collect())
+        .unwrap_or_else(|| std::iter::repeat(None).take(keys.len()).collect());
+
+    let representations = body
+        .variables
+        .get_mut(REPRESENTATIONS)
+        .and_then(|value| value.as_array_mut())
+        .expect("we already checked that representations exist");
+    // remove from representations the entities we already obtained from the cache
+    let (new_representations, cache_result, cache_control) =
+        filter_representations(&name, representations, keys, cache_result)?;
+
+    if let Some(control) = cache_control {
+        update_cache_control(&request.context, &control);
+    }
+
+    if !new_representations.is_empty() {
+        body.variables
+            .insert(REPRESENTATIONS, new_representations.into());
+
+        request
+            .context
+            .private_entries
+            .lock()
+            .insert(EntityCacheResults(cache_result));
+
+        Ok(ControlFlow::Continue(request))
+    } else {
+        let entities = cache_result
+            .into_iter()
+            .filter_map(|res| res.cache_entry)
+            .map(|entry| entry.data)
+            .collect::<Vec<_>>();
+        let mut data = Object::default();
+        data.insert(ENTITIES, entities.into());
+
+        Ok(ControlFlow::Break(
+            subgraph::Response::builder()
+                .data(data)
+                .extensions(Object::new())
+                .context(request.context)
+                .build(),
+        ))
+    }
+}
+
+fn update_cache_control(context: &Context, cache_control: &CacheControl) {
+    match context.private_entries.lock().get_mut::<CacheControl>() {
+        Some(c) => {
+            *c = c.merge(cache_control);
+        }
+        //FIXME: race condition. We need an Entry API for private entries
+        None => {
+            context.private_entries.lock().insert(cache_control.clone());
```

Comment:
Does it create a deadlock if we do this ?

Reply:
I don't know right away, so just in case, I updated it in [b529853](https://github.com/apollographql/router/pull/4170/commits/b5298535ac0b1754e975ae139831f462d467fe5b)